### PR TITLE
Troubleshoot compute-id

### PR DESCRIPTION
**Pull request by gibizer**
Depends-on: https://github.com/openstack-k8s-operators/edpm-ansible/pull/619

### CODE REVIEW

File: tests/roles/dataplane_adoption/tasks/main.yaml [hunk 1]
```diff
@@ -76,7 +76,7 @@
           ssh \
             -i {{ edpm_privatekey_path | default("/tmp/ansible_private_key") }} \
             {{ edpm_user }}@"${computes[$name]}" \
-            "grep -qF $uuid /var/lib/nova/compute_id || (echo $uuid | sudo tee /var/lib/nova/compute_id && sudo chown 42436:42436 /var/lib/nova/compute_id && sudo chcon -t container_file_t /var/lib/nova/compute_id)"
```
Comment (gibizer) [inline tests/roles/dataplane_adoption/tasks/main.yaml:79]: I remove the grep part to simplify things. If the file is already there with the correct uuid then writing it again should not cause any change.

Comment (gibizer) [reply tests/roles/dataplane_adoption/tasks/main.yaml:79]: I removed the container_t labeling of the file as in a greenfield deployment the file gets labeled with system_u:object_r:nova_var_lib_t:s0 instead.
```
[root@edpm-compute-0 ~]# stat /var/lib/nova/compute_id 
  File: /var/lib/nova/compute_id
  Size: 37        	Blocks: 8          IO Block: 4096   regular file
Device: fc01h/64513d	Inode: 20971652    Links: 1
Access: (0400/-r--------)  Uid: (42436/    nova)   Gid: (42436/    nova)
Context: system_u:object_r:nova_var_lib_t:s0
Access: 2024-03-19 15:59:03.534671483 +0000
Modify: 2024-03-19 15:58:57.582465452 +0000
Change: 2024-03-19 15:58:57.721470248 +0000
 Birth: 2024-03-19 15:58:57.582465452 +0000
```
Also added an extra stat of the file first to see the file attributes if the file is already there.

**Comment by gibizer** [comment]
recheck
pre adoption VM failed to start:
```
 'stderr_lines': ['#=#=#                                                                         ',
                  '',
                  '######################################################################## '
                  '100.0%##O#-#                                                                        ',
                  '##=O#- '
                  '#                                                                      ',
                  '',
                  '                                                                           '
                  '1.2%',
                  '######################################################################## '
                  '100.0%',
                  'Warning: Identity file '
                  '/home/zuul/install_yamls/out/edpm/ansibleee-ssh-key-id_rsa '
                  'not accessible: No such file or directory.',
                  'Failed to discover available identity versions when '
                  'contacting http://172.21.0.2:5000. Attempting to parse '
                  'version from URL.',
                  'Could not find versioned identity endpoints when attempting '
                  'to authenticate. Please check that your auth_url is '
                  'correct. Unable to establish connection to '
                  'http://172.21.0.2:5000: '
                  "HTTPConnectionPool(host='172.21.0.2', port=5000): Max "
                  'retries exceeded with url: / (Caused by '
                  "NewConnectionError('<urllib3.connection.HTTPConnection "
                  'object at 0x7f93db2cf700>: Failed to establish a new '
                  "connection: [Errno 111] Connection refused'))",
                  'Warning: Identity file '
                  '/home/zuul/install_yamls/out/edpm/ansibleee-ssh-key-id_rsa '
                  'not accessible: No such file or directory.',
                  'Failed to discover available identity versions when '
                  'contacting http://172.21.0.2:5000. Attempting to parse '
                  'version from URL.',
                  'Could not find versioned identity endpoints when attempting '
                  'to authenticate. Please check that your auth_url is '
                  'correct. Unable to establish connection to '
                  'http://172.21.0.2:5000: '
                  "HTTPConnectionPool(host='172.21.0.2', port=5000): Max "
                  'retries exceeded with url: / (Caused by '
                  "NewConnectionError('<urllib3.connection.HTTPConnection "
                  'object at 0x7f0d957b2490>: Failed to establish a new '
                  "connection: [Errno 111] Connection refused'))"],
```

**Comment by gibizer** [comment]
my bad
```
 'stderr_lines': ['+ declare -A computes',
                  '+ computes=(["standalone.localdomain"]="192.168.122.100")',
                  '+ for name in "${!computes[@]}"',
                  '++ oc exec -t openstackclient -- openstack hypervisor show '
                  'standalone.localdomain -f value -c id',
                  '+ uuid=0ba0860a-ab9b-48a8-b1fb-94bd5180705d',
                  "+ echo 'Writing 0ba0860a-ab9b-48a8-b1fb-94bd5180705d to "
                  "/var/lib/nova/compute_id on standalone.localdomain'",
                  '+ ssh -i '
                  '/home/zuul/src/github.com/openstack-k8s-operators/install_yamls//out/edpm/ansibleee-ssh-key-id_rsa '
                  "root@192.168.122.100 'stat /var/lib/nova/compute_id ; echo "
                  '0ba0860a-ab9b-48a8-b1fb-94bd5180705d | sudo tee '
                  '/var/lib/nova/compute_id && sudo chown 42436:42436 '
                  "/var/lib/nova/compute_id)'",
                  'Warning: Identity file '
                  '/home/zuul/src/github.com/openstack-k8s-operators/install_yamls//out/edpm/ansibleee-ssh-key-id_rsa '
                  'not accessible: No such file or directory.',
                  "bash: -c: line 1: syntax error near unexpected token `)'",
                  'bash: -c: line 1: `stat /var/lib/nova/compute_id ; echo '
                  '0ba0860a-ab9b-48a8-b1fb-94bd5180705d | sudo tee '
                  '/var/lib/nova/compute_id && sudo chown 42436:42436 '
                  "/var/lib/nova/compute_id)'"],
```

**Comment by gibizer** [comment]
recheck pre-launch VM failed, issue tracked in https://issues.redhat.com/browse/OSPCIX-227

**Comment by bogdando** [comment]
recheck RETRY_LIMIT

**Comment by gibizer** [comment]
recheck crc setup filed

**Comment by gibizer** [comment]
recheck

**Comment by gibizer** [comment]
recheck the adoption task timed out waiting for the first edpm deployment but it seem the deployment finished successfully when the must gather run. So I think we need to bump the wait time for the deployment during adoption

**Comment by fao89** [comment]
recheck

**Comment by gibizer** [comment]
recheck the issue is not reproduced:
```
{'changed': True,
 'cmd': 'set -euxo pipefail\n'
        '\n'
        '\n'
        'oc wait --for condition=Ready --timeout=300s Nova/nova\n',
 'delta': '0:00:00.124146',
 'end': '2024-04-04 07:55:14.639534',
 'msg': 'non-zero return code',
 'rc': 1,
 'start': '2024-04-04 07:55:14.515388',
 'stderr': '+ oc wait --for condition=Ready --timeout=300s Nova/nova\n'
           'Error from server (NotFound): nova.nova.openstack.org "nova" not '
           'found',
 'stderr_lines': ['+ oc wait --for condition=Ready --timeout=300s Nova/nova',
                  'Error from server (NotFound): nova.nova.openstack.org '
                  '"nova" not found'],
 'stdout': '',
 'stdout_lines': []}
 
```

**Comment by gibizer** [comment]
recheck issue is not reproduced
```
fatal: [localhost]: FAILED! => {"changed": true, "cmd": "set -euxo pipefail\n\n\noc wait --for condition=Ready --timeout=300s Nova/nova\n", "delta": "0:00:00.123971", "end": "2024-04-04 12:17:23.920450", "msg": "non-zero return code", "rc": 1, "start": "2024-04-04 12:17:23.796479", "stderr": "+ oc wait --for condition=Ready --timeout=300s Nova/nova\nError from server (NotFound): nova.nova.openstack.org \"nova\" not found", "stderr_lines": ["+ oc wait --for condition=Ready --timeout=300s Nova/nova", "Error from server (NotFound): nova.nova.openstack.org \"nova\" not found"], "stdout": "", "stdout_lines": []}
```
this not found error is strange. Nova/nova exists and ready according to https://logserver.rdoproject.org/70/370/773160e9ba239bb3547715a8e9a406dabf1a1203/github-check/data-plane-adoption-osp-17-to-extracted-crc-minimal-no-ceph/4ec60e5/controller/ci-framework-data/logs/openstack-k8s-operators-openstack-must-gather/namespaces/openstack/crs/nova.nova.openstack.org/nova.yaml

**Comment by gibizer** [comment]
recheck all green

**Comment by gibizer** [comment]
recheck

**Comment by gibizer** [comment]
recheck all green (total 4 green job runs)

**Comment by gibizer** [comment]
recheck all green (total 6)

**Comment by gibizer** [comment]
and 1 in 8 runs reproduced the issue again:
```
[1;35m/var/lib/nova/compute_id: operation not permitted[0m
[0;31mfatal: [standalone]: FAILED! => {"changed": false, "msg": "Failed containers: nova_compute"}[0m

**Comment by fao89** [comment]
probably we need a wait somewhere

**Comment by gibizer** [comment]
> probably we need a wait somewhere

I thought about this.
* the adoption side of the pre-generation of the file happens lot earlier than the container start as in between almost the whole 18 dataplane deployment runs
* the edpm_nova role during the 18 dataplane deployment also sets permission on the file and that happens 6 seconds before the container failure. In successful runs I see 11 and 9 seconds between the two events. So maybe this can be a place to introduce delay.

Also this issue is not visible in greenfield deployment but there we have different podman version:
* adoption 4.6.0.
* greenfield 4.9.4
but looking at the changelog between these podman version does not show any relevant change.

**Comment by gibizer** [comment]
recheck

**Comment by gibizer** [comment]
recheck CI is back after a day off

**Comment by gibizer** [comment]
recheck
minimal job timed out on the control plane deployment:
```
    message: OpenStackControlPlane Swift in progress
```

**Comment by gibizer** [comment]
> recheck
> 
> > > probably we need a wait somewhere
> > 
> > 
> > I thought about this.
> > 
> > * the adoption side of the pre-generation of the file happens lot earlier than the container start as in between almost the whole 18 dataplane deployment runs
> > * the edpm_nova role during the 18 dataplane deployment also sets permission on the file and that happens 6 seconds before the container failure. In successful runs I see 11 and 9 seconds between the two events. So maybe this can be a place to introduce delay.
> 
> Let's try with the edpm-ansible change that introduce 1 minute of pause after the permission change.

https://logserver.rdoproject.org/70/370/82d8d7f92733836cc9dfcaeeaf4678244fe5b21e/github-check/data-plane-adoption-osp-17-to-extracted-crc/0f9cd36/controller/ci-framework-data/logs/openstack-k8s-operators-openstack-must-gather/namespaces/openstack/pods/nova-compute-extraconfig-openstack-openstack-7425x/logs/openstackansibleee.log

so adding a 1 minute does not help so this is not a race around changing the permissions and starting a container.  

Next I will try to not adding +i to the file to see if that helps

**Comment by gibizer** [comment]
trying without the immutable flag on compute_id

**Comment by gibizer** [comment]
recheck (2 pass)

**Comment by gibizer** [comment]
recheck (4 passed)

**Comment by gibizer** [comment]
recheck (6 pass)

**Comment by gibizer** [comment]
recheck (7 pass)
we can ignore the retry limit as it did not reproduce the problem

**Comment by gibizer** [comment]
recheck (8 pass)

**Comment by gibizer** [comment]
recheck (10 pass)

**Comment by gibizer** [comment]
After 12 passed runs I conclude that the older podman version used in 17 (4.6.0) is buggy if the mounted file has the immutable file attribute set. While the podman version used in greenfield 18 (4.9.4) does not show the issue. So I will propose a change that will not add the immutable flag to stabilize adoption.

**Comment by gibizer** [comment]
Let's merge https://github.com/openstack-k8s-operators/edpm-ansible/pull/619 to stabilize adoption

**Comment by gibizer** [comment]
https://github.com/openstack-k8s-operators/edpm-ansible/pull/619 has been merged as  fix

**Pull request closed without merging**